README.md
# ChatBot

[![CI](https://github.com/deltachat-bot/chatbot/actions/workflows/python-ci.yml/badge.svg)](https://github.com/deltachat-bot/chatbot/actions/workflows/python-ci.yml)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

Conversational chat-bot for Delta Chat, using OpenAI API.

## Install

```sh
pip install git+https://github.com/deltachat-bot/chatbot.git
```

### Installing deltachat-rpc-server

This program depends on a standalone Delta Chat RPC server `deltachat-rpc-server` program that must be
available in your `PATH`. To install it check:
https://github.com/deltachat/deltachat-core-rust/tree/master/deltachat-rpc-server

## Usage

Configure the bot:

```sh
chatbot init bot@example.com PASSWORD
```

Create a configuration file named `config.json` in the program folder, for example:

```
{
  "api_key": "sk-...",
  "global_monthly_quota": 10000000,
  "user_hourly_tokens_quota": 4000,
  "user_hourly_queries_quota": 60,
  "openai": {
    "model": "gpt-3.5-turbo",
    "max_tokens": 500,
    "messages": [{"role": "system", "content": "You are a helpful assistant inside Delta Chat messenger."}],
    "temperature": 0.7
  }
}
```

**Note:** On GNU/Linux the program folder is at `~/.config/chatbot/`, when configuring the bot it will
print the path to the accounts folder that is inside the configuration folder.

Start the bot:

```sh
chatbot serve
```

Run `chatbot --help` to see all available options.

To change log level set the enviroment variable `CHATBOT_LOG_LEVEL` to one of:
`debug`, `info`, `warning` or `error`


## Using the bot in groups

To use the bot in groups just add it to a group and send a message starting with @BotName to
mention it or quote-reply a previous message from the the bot.


CHANGELOG.md
# Changelog

## Unreleased

- initial release


deltachat_chatbot/orm.py
"""database"""
from typing import Any

from sqlalchemy import Column, Integer
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import declarative_base, sessionmaker
from sqlalchemy.sql.selectable import Select

Base = declarative_base()
_session = None  # noqa


class Usage(Base):  # noqa
    """User quota stats"""

    __tablename__ = "usage"
    user_id = Column(Integer, primary_key=True)
    tokens = Column(Integer, nullable=False)
    queries = Column(Integer, nullable=False)
    ends_at = Column(Integer, nullable=False)

    def __init__(self, **kwargs) -> None:
        kwargs.setdefault("tokens", 0)
        kwargs.setdefault("queries", 0)
        super().__init__(**kwargs)


def async_session():
    """Get session"""
    return _session()


async def init(path: str, debug: bool = False) -> None:
    """Initialize engine."""
    global _session  # noqa
    engine = create_async_engine(path, echo=debug)
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    _session = sessionmaker(engine, class_=AsyncSession)


async def fetchone(session: sessionmaker, stmt: Select) -> Any:
    return (await session.execute(stmt.limit(1))).scalars().first()


deltachat_chatbot/quota.py
"""Cooldown loop logic"""
import asyncio
import time
from datetime import datetime, timedelta

from deltabot_cli import BotCli
from sqlalchemy.future import select
from sqlalchemy.sql.expression import delete

from .orm import Usage, async_session, fetchone


class QuotaManager:
    """Manage user and global quotas"""

    def __init__(self, cli: BotCli, cfg: dict) -> None:
        self._cli = cli
        self._cfg = cfg
        self._rate_limit = 0
        self._lock = asyncio.Lock()

    async def cooldown_loop(self) -> None:
        next_month = await self.get_global_cooldown()
        while True:
            if self._rate_limit <= time.time():
                self._rate_limit = 0
            if next_month <= time.time():
                async with self._lock:
                    self._cli.set_custom_config("used_tokens", "0")
                next_month = _get_next_month_timestamp()
                await self._cli.set_custom_config("next_month", str(next_month))

            async with async_session() as session:
                async with session.begin():
                    stmt = delete(Usage).filter(Usage.ends_at <= time.time())
                    await session.execute(stmt)
            await asyncio.sleep(5)

    async def increase_usage(self, user_id: int, tokens: int) -> None:
        # increase global usage
        async with self._lock:
            used_tokens = int(await self._cli.get_custom_config("used_tokens") or 0)
            await self._cli.set_custom_config("used_tokens", str(used_tokens + tokens))

        # increase user usage
        async with async_session() as session:
            async with session.begin():
                stmt = select(Usage).filter_by(user_id=user_id)
                usage = await fetchone(session, stmt)
                if not usage:
                    usage = Usage(user_id=user_id, ends_at=_get_next_hour_timestamp())
                    session.add(usage)
                usage.queries += 1
                usage.tokens += tokens

    def is_rate_limited(self) -> bool:
        return bool(self._rate_limit)

    def set_rate_limit(self, seconds: int) -> None:
        self._rate_limit = int(time.time()) + seconds

    async def get_global_cooldown(self) -> int:
        return int(await self._cli.get_custom_config("next_month") or 0)

    async def global_quota_exceeded(self) -> bool:
        global_quota = int(self._cfg.get("global_monthly_quota") or 0)
        used_tokens = int(await self._cli.get_custom_config("used_tokens") or 0)
        if global_quota and used_tokens >= global_quota:
            return True
        return False

    async def quota_exceeded(self, user_id: int) -> int:
        """If quota was exceeded return the cooldown, otherwise return zero"""
        tokens_quota = int(self._cfg.get("user_hourly_tokens_quota") or 0)
        queries_quota = int(self._cfg.get("user_hourly_queries_quota") or 0)
        now = int(time.time())
        async with async_session() as session:
            async with session.begin():
                stmt = select(Usage).filter_by(user_id=user_id)
                usage = await fetchone(session, stmt)
                if usage:
                    if usage.tokens >= tokens_quota or usage.queries >= queries_quota:
                        return usage.ends_at - now
        return 0


def _get_next_month_timestamp() -> int:
    return int(
        (datetime.today().replace(day=25) + timedelta(days=7))
        .replace(day=1, hour=0, minute=0, second=0, microsecond=0)
        .timestamp()
    )


def _get_next_hour_timestamp() -> int:
    return int((datetime.today() + timedelta(hours=1)).timestamp())


deltachat_chatbot/utils.py
"""Utilities"""
import asyncio
import logging
import os
from typing import Coroutine

_background_tasks = set()


def run_in_background(coro: Coroutine) -> None:
    """Schedule the execution of a coroutine object in a spawn task, keeping a
    reference to the task to avoid it disappearing mid-execution due to GC.
    """
    task = asyncio.create_task(coro)
    _background_tasks.add(task)
    task.add_done_callback(_background_tasks.discard)


def human_time_duration(seconds: int) -> str:
    hours, seconds = divmod(int(seconds), 60 * 60)
    minutes, seconds = divmod(int(seconds), 60)
    return f"{hours:02d}:{minutes:02d}:{seconds:02d}"


def get_log_level() -> int:
    """Get log level from environment variables. Defaults to INFO if not set."""
    level = os.getenv("CHATBOT_LOG_LEVEL", "info").upper()
    return int(getattr(logging, level))


deltachat_chatbot/hooks.py
"""Event Hooks"""
# pylama:ignore=W0603
import asyncio
import json
import logging
import os
from argparse import Namespace
from typing import List, Tuple

import openai
import tiktoken
from deltabot_cli import AttrDict, Bot, BotCli, EventType, const, events

from .openai import get_reply, init_openai
from .orm import init as init_db
from .quota import QuotaManager
from .utils import get_log_level, human_time_duration, run_in_background

cli = BotCli("chatbot", get_log_level())
cfg: dict = {}
quota_manager = QuotaManager(cli, {})


@cli.on_init
async def on_init(bot: Bot, _args: Namespace) -> None:
    if not await bot.account.get_config("displayname"):
        await bot.account.set_config("displayname", "ChatBot")
        status = "I am a conversational Delta Chat bot, you can chat with me in private"
        await bot.account.set_config("selfstatus", status)


@cli.on_start
async def _on_start(bot: Bot, args: Namespace) -> None:
    global quota_manager  # pylint:disable=C0103
    path = os.path.join(args.config_dir, "config.json")
    if os.path.exists(path):
        with open(path, encoding="utf-8") as config:
            cfg.update(json.load(config))
    cfg["openai"] = {"model": "gpt-3.5-turbo", "n": 1, **(cfg.get("openai") or {})}
    api_key = cfg.get("api_key", "")
    assert api_key, "API key is not set"
    await init_openai(api_key, cfg["openai"])

    path = os.path.join(args.config_dir, "sqlite.db")
    await init_db(f"sqlite+aiosqlite:///{path}")

    quota_manager = QuotaManager(cli, cfg)
    run_in_background(quota_manager.cooldown_loop())
    logging.info(
        "Listening for messages at: %s", await bot.account.get_config("configured_addr")
    )


@cli.on(events.RawEvent)
async def log_event(event: AttrDict) -> None:
    if event.type == EventType.INFO:
        logging.info(event.msg)
    elif event.type == EventType.WARNING:
        logging.warning(event.msg)
    elif event.type == EventType.ERROR:
        logging.error(event.msg)


@cli.on(events.MemberListChanged(added=True))
async def _member_added(event: AttrDict) -> None:
    msg = event.message_snapshot
    account = msg.message.account
    if event.member == await account.get_config("configured_addr"):
        await msg.chat.send_text("👋")


@cli.on(events.NewMessage(is_info=False, func=cli.is_not_known_command))
async def _filter_messages(event: AttrDict) -> None:
    msg = event.message_snapshot
    chat = await msg.chat.get_basic_snapshot()
    if not msg.text or not await _should_reply(msg, chat):
        return

    messages, prompt_tokens = await _get_messages(msg)
    if not messages:
        await msg.chat.send_message(text="TL;DR", quoted_msg=msg.id)
    else:
        global_quota_exceeded = await quota_manager.global_quota_exceeded()
        if global_quota_exceeded:
            cooldown = human_time_duration(await quota_manager.get_global_cooldown())
            await msg.chat.send_message(
                text=f"Quota exceeded, wait for: ⏰ {cooldown}", quoted_msg=msg.id
            )
            return

        quota_exceeded = await quota_manager.quota_exceeded(msg.from_id)
        if quota_exceeded > 0:
            cooldown = human_time_duration(quota_exceeded)
            await msg.chat.send_message(
                text=f"Quota exceeded, wait for: ⏰ {cooldown}", quoted_msg=msg.id
            )
            return

        if quota_manager.is_rate_limited():
            await msg.chat.send_message(
                text="⏰ I'm not available right now, try again later", quoted_msg=msg.id
            )
            return

        try:
            max_tokens = int(cfg["openai"].get("max_tokens") or 0)
            reply = await get_reply(
                str(msg.from_id), messages, max_tokens - prompt_tokens
            )
            logging.debug("bot reply: %s", reply)
            await quota_manager.increase_usage(msg.from_id, reply.usage.total_tokens)
            text = reply.choices[0].message.content.strip()
            await msg.chat.send_message(text=text, quoted_msg=msg.id)
            await asyncio.sleep(1)  # avoid rate limits
        except openai.error.RateLimitError:
            quota_manager.set_rate_limit(60)


async def _get_messages(msg: AttrDict) -> Tuple[List[dict], int]:
    text = ""
    if msg.quote and msg.quote.text:
        text = "> " + msg.quote.text.replace("\n", "\n> ") + "\n\n"

    max_tokens = int(cfg["openai"].get("max_tokens") or 0)
    prompt_tokens = 0
    if max_tokens:
        enc = tiktoken.encoding_for_model(cfg["openai"].get("model"))
        for text2 in (text + msg.text, msg.text):
            tokens = len(enc.encode(text2))
            if tokens <= max_tokens // 2:
                prompt_tokens = tokens
                text = text2
                break
        else:
            text = ""

    return [{"role": "user", "content": text}] if text else [], prompt_tokens


async def _should_reply(msg: AttrDict, chat: AttrDict) -> bool:
    # 1:1 direct chat
    if chat.chat_type == const.ChatType.SINGLE:
        return True

    # mentions
    account = msg.message.account
    selfaddr = await account.get_config("configured_addr")
    displayname = await account.get_config("displayname")
    mention = displayname and msg.text.startswith(f"@{displayname}")
    if msg.text.startswith(selfaddr) or mention:
        return True

    # quote-reply
    if msg.quote and msg.quote.get("message_id"):
        quote = account.get_message_by_id(msg.quote.message_id)
        snapshot = await quote.get_snapshot()
        if snapshot.sender == account.self_contact:
            return True

    return False
